pkg/image/decompress.go

```diff
@@ -2,14 +2,94 @@ package image
 
 import (
 	"archive/tar"
-	"fmt"
+	"compress/gzip"
 	"io"
-	"os"
 	"path/filepath"
+	"strings"
+
+	"github.com/golang/glog"
 )
 
-func IsCompressed(srcFile string) bool {
-	fExt := filepath.Ext(srcFile)
+var SupportedCompressionExtensions = []string{
+	Gz,
+}
+
+var SupportedArchiveExtentions = []string{
+	TarArch, TarArch + Gz,
+}
+
+// TODO fixup comment
+// UnpackData combines the decompressing and unarchiving of a data stream and returns the
+// end of the stream for further processing.
+func UnpackData(filename string, src io.Reader) interface{} {
+	glog.Infof("UnpackData: filename: %v, src: %T\n", filename, src)
+	/*if file is raw image{
+		return io.ReadCloser (raw image stream)
+	}*/
+	src = DecompressData(filename, src).(io.ReadCloser)
+	src = DearchiveData(filename, src).(io.Reader)
+	/*if file is qcow2 {
+	convert to raw image
+	return io.ReadCloser
+	}*/
+	return src
+}
+
+// TODO fixup comment
+// DecompressData analyzes the filename (string) to decide which decompression function to call.
+// Compression packages return objects that must be closed after reading,
+// hence the need to return a ReadCloser.  It is up to the caller of DecompressData
+// to close the returned stream.
+// If not compression is detected, it is considered a 'noop' and the original stream is
+// returned.
+func DecompressData(filename string, src io.Reader) interface{} {
+	glog.Infof("DecompressData: filename: %v, src: %T", filename, src)
+	if src == nil || filename == "" {
+		glog.Errorln("DecompressData: Nil parameter found.")
+		return nil
+	}
+	fn := filepath.Ext(strings.ToLower(strings.TrimSpace(filename)))
+	glog.Infof("DecompressData: checking ext: %v", fn)
+	switch fn {
+	case Gz:
+		glog.Infof("DecompressData: detected %v is of %v format", fn, Gz)
+		src = gunzip(src)
+	default:
+		// noop
+	}
+	return src
+}
+
+// TODO fixup comment
+// DearchiveData analyzes a filename (string) to decided which de-archive function to call.
+// Golang archive packages return readers and do not need to be closed, so only an io.Reader
+// is returned.
+// If not archive format is detected, it is considered a 'noop' and the original stream is
+// returned.
+func DearchiveData(filename string, src io.Reader) interface{} {
+	glog.Infoln("DearchiveData: filename: %v, src: %v", )
+	switch {
+	case strings.HasSuffix(filename, TarArch), strings.HasSuffix(filename, TarArch+Gz):
```

Comment:
For future additions probably good to pull out the file extension before the `case`

Reply:
That runs into trouble with nested extentions (.tar.gz) since `filepath.Ext()` will splice everything after the final `.`